File: Day 0009/Soluition.java
import java.util.*;

class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        // Count frequency of each element
        Map<Integer, Integer> mp = new HashMap<>();
        for (int num : nums)
            mp.put(num, mp.getOrDefault(num, 0) + 1);

        // Max heap based on frequency (value of map entries)
        PriorityQueue<Map.Entry<Integer, Integer>> maxHeap =
            new PriorityQueue<>((a, b) -> b.getValue() - a.getValue());

        // Add all frequency entries to the max heap
        maxHeap.addAll(mp.entrySet());

        // Extract top k frequent elements
        int[] result = new int[k];
        for (int i = 0; i < k; i++) {
            result[i] = maxHeap.poll().getKey();
        }

        return result;
    }
}

File: Problem002/Soluition.cpp
#include <vector>
#include <unordered_map>
#include <queue>
#include <utility> // for std::pair

using namespace std;

using pii = pair<int, int>; // Define a pair type for (element, frequency)
using vp = vector<pii>;     // Define a vector type for pairs

class Solution
{
public:
    vector<int> topKFrequent(vector<int> &nums, int k)
    {
        // Custom comparator for the max heap based on frequency
        auto comp = [](const pii &a, const pii &b) -> bool
        {
            return a.second < b.second; // We want a max heap, so we compare frequencies
        };

        // Priority queue (max heap) to store pairs of (element, frequency)
        priority_queue<pii, vp, decltype(comp)> MaxHeap(comp);
        unordered_map<int, int> mp; // Map to count frequencies of elements

        // Count the frequency of each element in nums
        for (auto &it : nums)
        {
            ++mp[it]; // Increment the count for each element
        }

        // Push all elements and their frequencies into the max heap
        for (const auto &it : mp)
        {
            MaxHeap.push(it); // Push the pair (element, frequency) into the heap
        }

        vector<int> res; // Result vector to store the top K frequent elements

        // Extract the top K elements from the max heap
        while (k-- && !MaxHeap.empty())
        {
            res.push_back(MaxHeap.top().first); // Get the element with the highest frequency
            MaxHeap.pop();                      // Remove it from the heap
        }

        return res; // Return the result
    }
};
